frontend/src/services/promotions.ts

```diff
@@ -1,98 +1,119 @@
-  DiscountType,
-  ServiceOptions,
-  Sort,
-} from "../types/promotion";
-import { Promotion } from "../types/promotion";
-import { enumContainsValue } from "../utils/enum";
+import { FilterOptions, Sort, Promotion } from "../types/promotion";
 
-export async function get(): Promise<Promotion[]> {
+/**
+ * Fetches list of promotions and sets them on this instance.
+ * If an error occurs, an empty list will be set on this instance.
+ */
+export async function getPromotions(): Promise<Promotion[]> {
   return fetch("/promotions")
     .then((res: Response) => res.json())
+    .then((promotions: Promotion[]) => promotions)
     .catch(() => []);
 }
 
-export function sortPromotions(
-  promotions: Promotion[],
-  sort: SortBy
+/**
+ * Returns the subset of this instance's promotions which satisfy at least one filter key in the `options` parameter.
+ *
+ * @param arr - The list of promotions to filter through
+ * @param options - An object specifying the keys and the values to filter the promotions by
+ */
+export function filterPromotions(
+  arr: Promotion[],
+  filters: FilterOptions
 ): Promotion[] {
-  switch (sort) {
-    case Sort.Distance:
-      return sortDistance(promotions);
-    case Sort.MostPopular:
-      return sortPopularity(promotions);
-    case Sort.Rating:
-      return sortRating(promotions);
-    default:
-      return promotions;
-  }
-}
+  const { cuisineType, dayOfWeek, discountType, serviceOptions } = filters;
 
-function sortDistance(promotions: Promotion[]) {
-  // TODO: sort based on restaurant distance
+  let promotions = [...arr];
+  if (filters.cuisineType.length > 0) {
+    promotions = filterCuisineType(promotions, cuisineType);
+  }
+  if (filters.dayOfWeek.length > 0) {
+    promotions = filterDayOfWeek(promotions, dayOfWeek);
+  }
+  if (filters.discountType.length > 0) {
+    promotions = filterDiscountType(promotions, discountType);
+  }
+  if (filters.serviceOptions.length > 0) {
+    promotions = filterServiceOptions(promotions, serviceOptions);
+  }
   return promotions;
 }
 
-function sortPopularity(promotions: Promotion[]) {
-  // TODO: sort based on number of saves
-  return promotions;
+function filterCuisineType(promotions: Promotion[], filters: string[]) {
+  let result: Promotion[] = [];
+  for (const key of filters) {
+    const filtered = promotions.filter(({ cuisine }) => {
+      const sanitized = cuisine.replace("\\s", "_");
+      return sanitized.toUpperCase() === key.toUpperCase();
+    });
+    result = [...result, ...filtered];
+  }
+  return result;
 }
 
-function sortRating(promotions: Promotion[]) {
-  // TODO: sort based on number of ratings
-  return promotions;
+function filterDayOfWeek(promotions: Promotion[], filters: string[]) {
+  let result: Promotion[] = [];
+  for (const key of filters) {
+    const filtered = promotions.filter(({ schedules }) => {
+      return schedules.find(({ dayOfWeek }) => dayOfWeek === key);
+    });
+    result = [...result, ...filtered];
```

Comment:
I didn't think about this! But I don't think it does... this is basically just appending it to the end of the result array. Does the BE query functionality remove duplicates?

Reply:
I think it might contain duplicates.  Like, say we have Promotion A which has two schedules Monday and Tuesday. Then the user wants to find all promotions for Monday and Tuesday. I think what happens is that the first iteration would add promotion A, the second iteration would append it again. 

Like for example something like this:
![image](https://user-images.githubusercontent.com/49849754/103466674-cfb24080-4cfb-11eb-997a-153f81f45d70.png)

> Does the BE query functionality remove duplicates?

Hmm we currently don't support any filtering for schedules, but I believe when we add it, we would prevent duplicates yah. I can start working on that now actually